### DISCUSSION TITLE
Thanks a lot

### DISCUSSION THREAD

**Discussion by goyalyashpal** [pcottle/learnGitBranching#1082] [General]
Hi!

thanks a lot for making this soooo well. 

now due to such awesome lessons - _backed by_ play area of sandbox [`?NODEMO`](https://learngitbranching.js.org?NODEMO), 
_BOTH_ of which have instant visualisation of what a command is doing, 

i tried all the commands i was doubtful at 3-6 times in the sandbox, and experimented until i was confident about the outcome of a command - and that yeah, i am not just guessing about it anymore.

I never dared to rebase before. thereafter, i have already done _meaningful_ rebase 3-4 times in last 10 hours. 
same goes for creating branch, meaning of the checkout-HEAD/branch, etc ...

i knew oh-my-git from years ago, but got intimidated on its install procedures on windows;
just found learngitbranching yesterday via [alternativeto.net](https://alternativeto.net/software/oh-my-git-/) - thanks for keeping the entry barrier extremely low.

its awesome. thanks a lot.

----
edit: oops, sorry https://github.com/pcottle/learnGitBranching/issues/498

**Comment by pcottle** [comment]
Glad this was helpful to you @goyalyashpal ! I felt a similar way -- I never understood what a rebase was until I saw a visual demonstration. That's what inspired me to make this whole project 😇

**Reply by goyalyashpal** [reply]
> felt a similar way -- I never understood what a rebase was until I saw a visual demonstration.

on the contrary, i had seen several visualisations already; but even those didn't help as they were passive videos (or articles)
it never settled in memory as there was no playground, no sandbox - to help induce the sense of doubt clearing and hands on affirmation; for the various "what-ifs" , or "how-tos" or "can-this" which raise in mind while trying to do such a tangible yet opaque-ish task.

so, those passive floating visualisations; despite being on the spot and well made too; faded as quickly

whereas **_here_** on your this platform, it is an all perfect receipe - contains all the elements

0. no installations, no hassles
1. first, a passive introduction; say: [`level intro4; help level`](https://learngitbranching.js.org/?NODEMO&command=level%20intro4;help%20level)
2. then a planned exercise 
3. ... with terminal commands _(and not abstracted away dumbed down buttons like those gooey/GUI IDEs)_
4. ... and with visualisations;
5. then the sandbox playground to go wild and self validate
6. ... all as much to mind's content without any paywall

**Reply by goyalyashpal** [reply]
just today too, i had to accomplish a git task - which for me is fairly complicated. 
but with learngitbranching sandbox, i visualised & tried it there to be safe, then carried over to actual git :)

~input for pasting in lgb prompt:~ edit: just re-learned that [semi-colon in bash](https://duckduckgo.com/?q=semi-colon+in+bash&t=vivaldi&ia=web&iax=qa) is valid bash too :)

```
git commit;
git commit -m "wrong commit";
git commit;
git commit;

# Task sequence starts here;
git checkout HEAD~3 -b amend;
git commit --amend;
git rebase amend main;
git branch -d amend
```

it's nice that pasting it there in this format executes step by step - makes it awesome for such record keeping of illustrations for techniques as well :)

it's such a bliss for this tool to exist. god bless u. (sorry if u're atheist lol 😅)

**Reply by goyalyashpal** [reply]
also, did i do that right? or is there some better way to do this too??

**Reply by pcottle** [reply]
You can make little demonstration links by just adding commands onto the URL!

https://learngitbranching.js.org/?NODEMO&command=gc;git%20branch%20badCommit;gc;gc;gc;git%20checkout%20HEAD~3;%20git%20commit%20--amend;%20git%20branch%20amend;%20git%20rebase%20amend%20main;%20git%20branch%20-d%20amend;

**Reply by pcottle** [reply]
And yeah that's a fine sequence to fix this mistake!

**Reply by goyalyashpal** [reply]
Just pasting the useful seeming output from git here for later reference:
(as it mentions the new `git switch` thingy)

```
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false
```
